src/index.js

```diff
@@ -0,0 +1,159 @@
+import React, { PureComponent } from 'react';
+import { render } from 'react-dom';
+import PropTypes from 'prop-types';
+import { IntlProvider, addLocaleData, FormattedMessage } from 'react-intl';
+import NotInitialisedError from './NotInitialisedError';
+import supportedLanguages from './supportedLanguages.json';
+
+let translate = NotInitialisedError;
+let formatDate = NotInitialisedError;
+let Translation = NotInitialisedError;
+
+const withI18n = Component => {
+  class WithI18nComponent extends PureComponent {
+    render() {
+      return <Component {...this.props} translate={translate} formatDate={formatDate} />;
+    }
+  }
+
+  return WithI18nComponent;
+};
+
+class Provider extends PureComponent {
+  state = {
+    loaded: false,
+  };
+
+  async componentDidMount() {
+    const { domain } = this.props;
+    const language = this.getUserLanguage();
+    const localeData = this.getAllLocaleData();
+
+    addLocaleData(localeData);
+    const translations = await this.fetchTranslations(language);
+
+    const component = React.createElement(IntlProvider, {
+      messages: translations,
+      children: React.createElement('div'),
+      locale: language,
+      ref: element => {
+        const { intl } = element.getChildContext();
+
+        const provideIntlContext = Component => {
+          class IntlComponent extends PureComponent {
+            static childContextTypes = {
+              intl: PropTypes.object,
+            };
+
+            static propTypes = {
+              id: PropTypes.string.isRequired,
+            };
+
+            getChildContext() {
+              return { intl };
+            }
+
+            render() {
+              const { id, ...others } = this.props;
+              return <Component {...others} id={`${domain}.${id}`} />;
+            }
+          }
+
+          return IntlComponent;
+        };
+
+        translate = (id, values) => intl.formatMessage({ id: domain ? `${domain}.${id}` : id }, values);
+        formatDate = intl.formatDate;
+        Translation = provideIntlContext(FormattedMessage);
+
+        this.setState({ loaded: true });
+      },
+    });
+
+    render(component, document.createElement('div'));
+  }
+
+  getUserLanguage() {
+    if (this.props.isDevelopmentMode) {
+      return this.props.developmentLanguage;
+    }
+
+    const language = document.body.parentNode.getAttribute('lang');
```

Comment:
PP: `document.documentElement.lang`, access the real root and not just the parent of the `<body` (very edge case, I know)